// kodi-jsonrpc-http-post/index.js
'use strict';

const btoa = require('btoa');
const request = require('request');

module.exports = {
    clientIp: null,
    clientPort: null,
    authorizationBasic: null,

    init: function(clientIp, clientPort, clientId, clientSecret) {
        var _self = this;
        return new Promise(function(resolve, reject){
            try {
                // Turn login data to Base64 encoded ascii:
                _self.authorizationBasic = btoa(clientId + ':' + clientSecret);
                
                _self.clientIp      = clientIp;
                _self.clientPort    = clientPort;
                resolve("kodi init complete");
            } catch (err) {
                reject("Error initializing kodi " + err);
            }
        });
    },
    sendCommand:function(postData){
        var _self = this;
        
        return new Promise(function(resolve, reject){
            if( _self.clientIp == null || _self.clientPort == null || _self.authorizationBasic == null) {
                reject(Error("Needed variables missing, make sure init has been run"));
            }
            else {
                // Combine our URL:
                var url = 'http://' + _self.clientIp + ":" + _self.clientPort + "/jsonrpc";
                // Setup our request options:
                var options = {
                    method: 'post',
                    headers : {
                        "Content-Type" : "application/json",
                        "Authorization" : 'Basic ' + _self.authorizationBasic
                    },
                    // Our sent in json data request:
                    body: postData,
                    json: true,
                    url: url
                }
                // Run our request:
                request(options, function (err, res, body) {
                    if (err) {
                        reject('error posting json: ' + err);
                    }
                    else {
                        // console.log('headers: ', res.headers)
                        // console.log('statusCode: ', res.statusCode)
                        resolve(body);
                    }
                });
            }
        });
    },
}
